packages/@aws-cdk/core/lib/cfn-parse.ts

```diff
@@ -682,48 +682,72 @@ export class CfnParser {
       : undefined;
   }
 
-  private parseFnSubString(value: string, map: { [key: string]: any } = {}): string {
-    const leftBrace = value.indexOf('${');
-    const rightBrace = value.indexOf('}') + 1;
-    // don't include left and right braces when searching for the target of the reference
-    if (leftBrace === -1 || leftBrace >= rightBrace) {
-      return value;
-    }
-
-    const leftHalf = value.substring(0, leftBrace);
-    const rightHalf = value.substring(rightBrace);
-    const refTarget = value.substring(leftBrace + 2, rightBrace - 1).trim();
-    if (refTarget[0] === '!') {
-      return value.substring(0, rightBrace) + this.parseFnSubString(rightHalf, map);
-    }
-
-    // lookup in map
-    if (refTarget in map) {
-      return leftHalf + '${' + refTarget + '}' + this.parseFnSubString(rightHalf, map);
-    }
-
-    // since it's not in the map, check if it's a pseudo parameter
-    const specialRef = this.specialCaseSubRefs(refTarget);
-    if (specialRef !== undefined) {
-      return leftHalf + specialRef + this.parseFnSubString(rightHalf, map);
-    }
+  private parseFnSubString(templateString: string, expressionMap: { [key: string]: any } | undefined): string {
+    const map = expressionMap ?? {};
+    const self = this;
+    return Fn.sub(go(templateString), Object.keys(map).length === 0 ? expressionMap : map);
+
+    function go(value: string): string {
+      const leftBrace = value.indexOf('${');
+      const rightBrace = value.indexOf('}') + 1;
+      // don't include left and right braces when searching for the target of the reference
+      if (leftBrace === -1 || leftBrace >= rightBrace) {
+        return value;
+      }
+
+      const leftHalf = value.substring(0, leftBrace);
+      const rightHalf = value.substring(rightBrace);
+      const refTarget = value.substring(leftBrace + 2, rightBrace - 1).trim();
+      if (refTarget[0] === '!') {
+        return value.substring(0, rightBrace) + go(rightHalf);
+      }
+
+      // lookup in map
+      if (refTarget in map) {
+        return leftHalf + '${' + refTarget + '}' + go(rightHalf);
+      }
+
+      // since it's not in the map, check if it's a pseudo-parameter
+      // (or a value to be substituted for a Parameter, provided by the customer)
+      const specialRef = self.specialCaseSubRefs(refTarget);
```

Comment:
Are pseudo-parameters even supported if there is an expression map?

I think as soon as there is a map, every key has to come from the map, no?

Reply:
No. There is no such limitation. This is valid CloudFormation:

```yaml
Resources:
  SNSTopic:
    Type: AWS::SNS::Topic
    Properties:
      TopicName:
        Fn::Sub:
          - my-sns-${Param}-${AWS::Region}-topic-${Environment}
          - Environment:
              Ref: AnotherParam
```

(where `Param` and `AnotherParam` are Parameters of the template)